File: examples/java/OauthPkce.java
// 生成 PKCE 与授权 URL。Java 11+
// java OauthPkce

import java.net.URLEncoder;
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.SecureRandom;
import java.util.Base64;

public class OauthPkce {
    public static void main(String[] args) throws Exception {
        String verifier = b64url(random(32));
        MessageDigest md = MessageDigest.getInstance("SHA-256");
        String challenge = b64url(md.digest(verifier.getBytes(StandardCharsets.US_ASCII)));
        String state = b64url(random(16));

        String instance = env("INSTANCE", "https://example.com").replaceAll("/+$", "");
        String clientId = env("CLIENT_ID", "your-client-id");
        String redirect = env("REDIRECT_URI", "https://your-app.example/oauth/callback");
        String scope = env("SCOPE", "read:profile write:notes");

        String url = instance + "/oauth/authorize"
                + "?response_type=code"
                + "&client_id=" + enc(clientId)
                + "&redirect_uri=" + enc(redirect)
                + "&scope=" + enc(scope)
                + "&state=" + enc(state)
                + "&code_challenge=" + enc(challenge)
                + "&code_challenge_method=S256";

        System.out.println("{");
        System.out.println("  \"code_verifier\": \"" + verifier + "\",");
        System.out.println("  \"code_challenge\": \"" + challenge + "\",");
        System.out.println("  \"state\": \"" + state + "\",");
        System.out.println("  \"authorize_url\": \"" + url + "\"");
        System.out.println("}");
    }

    static byte[] random(int n) {
        byte[] b = new byte[n];
        new SecureRandom().nextBytes(b);
        return b;
    }

    static String b64url(byte[] b) {
        return Base64.getUrlEncoder().withoutPadding().encodeToString(b);
    }

    static String enc(String s) {
        return URLEncoder.encode(s, StandardCharsets.UTF_8);
    }

    static String env(String k, String def) {
        String v = System.getenv(k);
        return v == null || v.isEmpty() ? def : v;
    }
}

File: examples/go/oauth_pkce.go
// 生成 PKCE 与授权 URL。
// 运行: go run oauth_pkce.go
package main

import (
	"crypto/rand"
	"crypto/sha256"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"net/url"
	"os"
)

func main() {
	verifier := b64url(randBytes(32))
	sum := sha256.Sum256([]byte(verifier))
	challenge := b64url(sum[:])
	state := b64url(randBytes(16))

	instance := env("INSTANCE", "https://example.com")
	clientID := env("CLIENT_ID", "your-client-id")
	redirect := env("REDIRECT_URI", "https://your-app.example/oauth/callback")
	scope := env("SCOPE", "read:profile write:notes")

	q := url.Values{}
	q.Set("response_type", "code")
	q.Set("client_id", clientID)
	q.Set("redirect_uri", redirect)
	q.Set("scope", scope)
	q.Set("state", state)
	q.Set("code_challenge", challenge)
	q.Set("code_challenge_method", "S256")

	out, _ := json.MarshalIndent(map[string]string{
		"code_verifier":  verifier,
		"code_challenge": challenge,
		"state":          state,
		"authorize_url":  instance + "/oauth/authorize?" + q.Encode(),
	}, "", "  ")
	fmt.Println(string(out))
}

func randBytes(n int) []byte {
	b := make([]byte, n)
	_, _ = rand.Read(b)
	return b
}

func b64url(b []byte) string {
	return base64.RawURLEncoding.EncodeToString(b)
}

func env(k, def string) string {
	if v := os.Getenv(k); v != "" {
		return trimSlash(v)
	}
	return trimSlash(def)
}

func trimSlash(s string) string {
	for len(s) > 0 && s[len(s)-1] == '/' {
		s = s[:len(s)-1]
	}
	return s
}
